### PR TITLE
Adding FastAI(pytorch) as a first class citizen to the pipeline

### PR DESCRIPTION
**Pull request by adricl**
This PR allows you to use Fast AI(pytorch) end to end in donkey car. I have tested training and driving with the simulator.

I have added a new type of model fastai_ so far linear is the only model implemented. More can be added in the future of course.
I have tried my best to emulate most of the function that Keras does but there are some idiosyncrasies between the two frameworks. 

These normal training and driving commands work.
python train.py --tubs data/ --model models/fastai_linear_model.pth --type fastai_linear
python manage.py drive --model models/fastai_linear_model.pth --type fastai_linear

I do realise that this PR comes without a heads up or warning so if its rejected that's fine. 
I do really like this project and the work that has been done, its just I prefer PyTorch to TensorFlow. 

Any suggestions are welcome. I am new to python of this complexity so any pointers would be great.
There will be more cleanups and fixes on this PR as I am not finished yet. I need to add some graph plotting and other tidy ups and maybe a refactor. Any guidance would be great. 
I will also add some docs around this.

If this goes well I might also look at getting pytorch lightning into own part and working in the pipeline.

Please note this will not work with TensorRT or TF lite if there is demand I could look at this in the future.

Cheers

### CODE REVIEW

**Comment by DocGarbanzo** [comment]
@adricl - are you still working on this PR? It will take a bit to go through this. You should definitely make sure that the tests are passing. If you are still working on it, then you can convert the PR to `draft`.

**Comment by adricl** [comment]
> Contributor

Thanks for that, I know its a big piece of work. Sorry about that. 
I am trying to get the tests to pass. I have converted it to draft,.

**Comment by tikurahul** [comment]
One more thing. When you are ready, can you please squash your changes to a single commit ? That way it becomes easier to reason about.

**Comment by adricl** [comment]
@tikurahul to squash the commit do I have to create a new PR or can I do it in this existing PR? I am not sure how to go about it if you can do it for the existing pr. Thanks

**Comment by adricl** [comment]
I think this pr is good to review now..

**Comment by DocGarbanzo** [comment]
@adricl - you don't need to do a new PR. You can squash the commits on your branch, make sure everything is working and force-push. It will update this PR here in-place.

File: donkeycar/parts/interpreter.py [hunk 1]
```diff
@@ -165,6 +169,70 @@ def load_weights(self, model_path: str, by_name: bool = True) -> \
...
+            self.model = torch.load(model_path)
+        else:
+            print("cuda not available for torch inference")
+            self.model = torch.load(model_path, map_location=torch.device('cpu'))
+
+        print(self.model)
+        self.model.eval()
+
+    def summary(self) -> str:
+        return self.model
```
Comment (DocGarbanzo) [inline donkeycar/parts/interpreter.py:235]: Does this return a string for description?

File: donkeycar/parts/interpreter.py [hunk 2]
```diff
@@ -165,6 +169,70 @@ def load_weights(self, model_path: str, by_name: bool = True) -> \
...
+
+    def predict_from_dict(self, input_dict):
+        for k, v in input_dict.items():
+            input_dict[k] = np.expand_dims(v, axis=0)
+        return self.invoke(input_dict)
+
+    def load(self, model_path: str) -> None:
+        logger.info(f'Loading model {model_path}')
+        if torch.cuda.is_available():
+            print("using cuda for torch inference")
```
Comment (DocGarbanzo) [inline donkeycar/parts/interpreter.py:225]: Pls use logging instead of print

File: donkeycar/parts/interpreter.py [hunk 3]
```diff
@@ -165,6 +169,70 @@ def load_weights(self, model_path: str, by_name: bool = True) -> \
...
+    def predict(self, img_arr: np.ndarray, other_arr: np.ndarray) \
+            -> Sequence[Union[float, np.ndarray]]:
+
+        inputs = torch.unsqueeze(img_arr, 0)
+        if other_arr is not None:
+            #other_arr = np.expand_dims(other_arr, axis=0)
+            inputs = [img_arr, other_arr]
+        return self.invoke(inputs)
+
+    def predict_from_dict(self, input_dict):
```
Comment (DocGarbanzo) [inline donkeycar/parts/interpreter.py:217]: Do you use this function?

File: donkeycar/parts/pytorch/torch_data.py [hunk 4]
```diff
@@ -41,7 +41,7 @@ def get_default_transform(for_video=False, for_inference=False):
     return transform
 
 
-class TorchTubDataset(IterableDataset):
+class FastaAiTorchTubDataset(IterableDataset):
```
Comment (DocGarbanzo) [inline donkeycar/parts/pytorch/torch_data.py:44]: Looks like this class has now been defined twice, pls check comment above.

File: donkeycar/pipeline/training.py [hunk 5]
```diff
@@ -81,6 +81,14 @@ def create_tf_data(self) -> tf.data.Dataset:
             output_shapes=self.model.output_shapes())
         return dataset.repeat().batch(self.batch_size)
 
+    def create_tf_data(self) -> tf.data.Dataset:
```
Comment (DocGarbanzo) [inline donkeycar/pipeline/training.py:84]: Please remove, this method is already defined 8 lines above.

File: donkeycar/pipeline/training.py [hunk 6]
```diff
@@ -162,12 +179,12 @@ def train(cfg: Config, tub_paths: str, model: str = None,
...
+        'History': history,
         'Transfer': os.path.basename(transfer) if transfer else None,
         'Comment': comment,
         'Config': str(cfg)
     }
     database.add_entry(database_entry)
     database.write()
 
-    return history
+    return history
```
Comment (DocGarbanzo) [inline donkeycar/pipeline/training.py:190]: EOF.

File: donkeycar/utils.py [hunk 7]
```diff
@@ -428,15 +428,17 @@ def eprint(*args, **kwargs):
     print(*args, file=sys.stderr, **kwargs)
 
 
-def get_model_by_type(model_type: str, cfg: 'Config') -> 'KerasPilot':
+def get_model_by_type(model_type: str, cfg: 'Config') -> 'ABC':
```
Comment (DocGarbanzo) [inline donkeycar/utils.py:431]: You can return `Union['KerasPilot', 'FastAiPilot']`

File: donkeycar/parts/interpreter.py [hunk 8]
```diff
@@ -3,15 +3,16 @@
 import logging
 import numpy as np
 from typing import Union, Sequence, List
+from pathlib import Path
```
Comment (DocGarbanzo) [inline donkeycar/parts/interpreter.py:6]: Are you using this?

File: donkeycar/utils.py [hunk 9]
```diff
@@ -428,15 +428,17 @@ def eprint(*args, **kwargs):
...
     '''
     given the string model_type and the configuration settings in cfg
     create a Keras model and return it.
     '''
     from donkeycar.parts.keras import KerasCategorical, KerasLinear, \
         KerasInferred, KerasIMU, KerasMemory, KerasBehavioral, KerasLocalizer, \
         KerasLSTM, Keras3D_CNN
-    from donkeycar.parts.interpreter import KerasInterpreter, TfLite, TensorRT
+    from donkeycar.parts.interpreter import KerasInterpreter, TfLite, TensorRT, FastAIInterpreter
```
Comment (DocGarbanzo) [inline donkeycar/utils.py:439]: Pls keep pep-8 compliant 80 char lines.

**Review by DocGarbanzo — CHANGES_REQUESTED** [review]
It's a great idea to bring in fastai. A couple of points need to be addressed. Please also add a test for the fastai model - you can compare how @EricWiener adapted the Keras tests to PyTorch. We should test the functionality of the model and the pipeline and the training. The latter should be switched off in the CI (we have the same setup also for TF and PyTorch).

**Comment by adricl** [comment]
Sorry I was squashing the commits and I Github closed the PR I have created a new one in #986 986 I will address your issues there. So sorr

File: donkeycar/parts/interpreter.py (hunk 1 above)
Comment (adricl) [reply donkeycar/parts/interpreter.py]: Yeah this does return a string, I have modified the Keras Class to return the same.

File: donkeycar/pipeline/training.py (hunk 5 above)
Comment (adricl) [reply donkeycar/pipeline/training.py]: So sorry about this.

File: donkeycar/utils.py (hunk 9 above)
Comment (adricl) [reply donkeycar/utils.py]: Fixed.

File: donkeycar/parts/interpreter.py (hunk 8 above)
Comment (adricl) [reply donkeycar/parts/interpreter.py]: No this was a mistake

File: donkeycar/utils.py (hunk 7 above)
Comment (adricl) [reply donkeycar/utils.py]: Done

File: donkeycar/parts/interpreter.py (hunk 3 above)
Comment (adricl) [reply donkeycar/parts/interpreter.py]: This has been removed

File: donkeycar/parts/interpreter.py (hunk 2 above)
Comment (adricl) [reply donkeycar/parts/interpreter.py]: Updated to use loggging.

File: donkeycar/parts/fastai.py [hunk 10]
```diff
@@ -1,31 +1,406 @@
...
+                                  patience=patience,
+                                  min_delta=min_delta),
+            SaveModelCallback(monitor='valid_loss',
+                              every_epoch=False
+                              )
+        ]
+
+        self.learner = Learner(dataLoader, model, loss_func=self.loss, path=Path(model_path).parent)
+
+        print(self.learner.summary())
```
[the remark this answers is not in the document]
Comment (adricl) [reply donkeycar/parts/fastai.py]: Updated

File: donkeycar/parts/fastai.py [hunk 11]
```diff
@@ -1,31 +1,406 @@
...
+        torch.save(self.learner.model, model_path)
+
+        if show_plot:
+            self.learner.recorder.plot_loss()
+            plt.savefig(Path(model_path).with_suffix('.png'))
+
+        history = list(map((lambda x: x.item()), self.learner.recorder.losses))
+
+        return history
+
```
[the remark this answers is not in the document]
Comment (adricl) [reply donkeycar/parts/fastai.py:278]: I have removed these functions.

File: donkeycar/parts/fastai.py [hunk 12]
```diff
@@ -1,31 +1,406 @@
...
+        self.optimizer = self.optimizer
+        self.loss = 'mse'
+
+    def interpreter_to_output(self, interpreter_out):
+        interpreter_out = (interpreter_out * 2) - 1
+        steering = interpreter_out[0]
+        throttle = interpreter_out[1]
+        print(steering, throttle)
         return steering, throttle
-
```
[the remark this answers is not in the document]
Comment (adricl) [reply donkeycar/parts/fastai.py:31]: I am not sure if these have been removed now. I am on linux so this is quite odd.

File: donkeycar/pipeline/training.py (hunk 6 above)
Comment (adricl) [reply donkeycar/pipeline/training.py:181]: Done

File: donkeycar/parts/fastai.py [hunk 13]
```diff
@@ -1,31 +1,406 @@
...
+from torchvision import transforms
+
+ONE_BYTE_SCALE = 1.0 / 255.0
+
+# type of x
+XY = Union[float, np.ndarray, Tuple[Union[float, np.ndarray], ...]]
+
+logger = getLogger(__name__)
+#TODO change this to use fastai Transforms
+def get_default_transform():
```
[the remark this answers is not in the document]
Comment (adricl) [reply donkeycar/parts/fastai.py]: I have fixed this so we don't have duplicate code. I did this as I thought the changes I would be doing were going to be huge, they turned out to be so minor.

In the future(not this PR) I will make use of some of the other transforms that fastai provides or albumentations if you guys implement it.

File: donkeycar/parts/pytorch/torch_data.py (hunk 4 above)
Comment (adricl) [reply donkeycar/parts/pytorch/torch_data.py]: This has been reverted.

**Comment by adricl** [comment]
I have to look at the tests next. That might take me a few days.

File: donkeycar/parts/fastai.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (DocGarbanzo) [reply donkeycar/parts/fastai.py:31]: Maybe I'm misinterpreting the git diff here. Should be all good I guess.

File: donkeycar/parts/interpreter.py (hunk 1 above)
Comment (DocGarbanzo) [reply donkeycar/parts/interpreter.py:229]: Hm... looking at lines 220 and 223, it doesn't look like `self.model` is string but rather a PyTorch model.

File: donkeycar/pipeline/training.py (hunk 6 above)
Comment (DocGarbanzo) [reply donkeycar/pipeline/training.py:182]: It still shows missing newline at EOF.

File: install/envs/ubuntu.yml [hunk 14]
```diff
@@ -36,6 +37,7 @@ dependencies:
   - plotly
   - pyyaml
   - tensorflow=2.2.0
+  - fastai
```
Comment (DocGarbanzo) [inline install/envs/ubuntu.yml:40]: Please also add the module in `setup.py`. Do you know if this installs on the RPi w/o problems?

**Comment by DocGarbanzo** [comment]
@adricl - this looks much better now.

File: install/envs/ubuntu.yml (hunk 14 above)
Comment (adricl) [reply install/envs/ubuntu.yml:40]: I think it should work as fastai is a wrapper around pytorch so there is no binaries as such. I will test this.

**Comment by Bleyddyn** [comment]
Thanks for this. I had just started working on the same idea but I haven't gotten nearly as far.
I have a notebook that has a working DataBlock for loading v2 Tub files, but I'm still trying to figure out how to train using it: https://github.com/Bleyddyn/malpi/tree/master/notebooks

**Comment by adricl** [comment]
@Bleyddyn off topic but this is how it was done in Tub V1 [link](https://github.com/cmasenas/fastai_navigation_training/blob/master/fastai_train.ipynb) This was a very clean implementation. I do have created a notebook so we can experiment with models but its not releasable as yet. I will have to rewrite it for this pr.

**Comment by DocGarbanzo** [comment]
@adricl @Bleyddyn - if you want to put records into a data frame as in the example above, this is now much easier with the new tub, because it is an Iterable over dictionaries. You just say:
```python
import pandas as pd
from donkeycar.parts.tub_v2 import Tub
t = Tub('/home/me/mycar/data')
df = pd.DataFrame(t)
```

**Comment by Bleyddyn** [comment]
@DocGarbanzo That's what I'm doing in the notebook I linked to. Updated now with a better DataBlock definition and actual working training code.

**Comment by DocGarbanzo** [comment]
@Bleyddyn - just looked at your notebook now. Very good.

**Comment by adricl** [comment]
@DocGarbanzo I have updated the test train pipeline to run fastai. I have also added fastai to the pi build. Is there anything else that needs to be completed for this PR?

File: donkeycar/tests/test_train.py [hunk 15]
```diff
@@ -13,7 +13,7 @@
 from donkeycar.utils import get_model_by_type, normalize_image, train_test_split
 
 Data = namedtuple('Data',
-                  ['type', 'name', 'convergence', 'pretrained', 'preprocess'],
+                  ['type', 'name', 'convergence', 'pretrained', 'preprocess', 'tf_lite', 'tensor_rt'],
                   defaults=(None, ) * 5)
```
Comment (DocGarbanzo) [inline donkeycar/tests/test_train.py:17]: If you add 2 more arguments to the namedtuple, please also increase the default initialisation by 2, i.e. 5 -> 7.

**Review by DocGarbanzo — CHANGES_REQUESTED** [review]
This is very good now, can you please address the small comment on the test & push it, then I'll merge the code.

File: donkeycar/tests/test_train.py (hunk 15 above)
Comment (adricl) [reply donkeycar/tests/test_train.py]: This is now resolved

**Comment by adricl** [comment]
We are good to go!

**Comment by c1505** [comment]
Awesome contribution here !  I would like to try this out in the simulator and/or on a car.  Any docs associated with this update or plan to create/modify some ? I can help review/validate the doc changes . @adricl

**Comment by adricl** [comment]
Thanks,
I will have to add documentation at some stage. 
Basically you need to train and load with model_type = "fastai_linear". This will will work end to end with the simulator as that is where I was running it. 
I will get onto the docs asap.

**Comment by adricl** [comment]
@c1505  this is my documentation. [FastAi Docs](https://github.com/adricl/donkeydocs/blob/master/docs/parts/fastai.md) Have a look before I merge it.

**Pull request merged**